### PR TITLE
Octos- Ari- oo-Ride-Share

### PR DESCRIPTION
**Pull request by arrriiii**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer | Instructor Feedback
:------------- | :------------- | ---
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I decided write an in_progress method on both passenger & driver (instead of doing one in_progress method in Trips for both classes). I figure the needs of the passenger may be different than the driver--and maybe passengers would like to be able to filter trips differently.
Describe any examples of composition that you encountered in this project, if any | Under the same module RideShare I was able to call methods/attributes from other classes in new methods such as duration and end_time. | Your answer gets at this, but I want to call out this vocab explicitly. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
Describe a _nominal test_ that you wrote for this assignment. | Calculating the passenger totals for total money spent, and duration have clear expected outcomes. -- Also, the driver total revenue and average revenue per hour.
Describe an _edge case test_ that you wrote for this assignment. | If I had more time, I would have written more edge case tests-- if the duration of a trip is shorter or longer than expected or when/if the trip cost is less than the fee.
Describe a concept that you gained more clarity on as you worked on this assignment. | I have a better understanding of dependencies and what it means to utilize subclasses as attributes for other classes.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -16,14 +18,17 @@ def initialize(input)
...
       @vehicle_id = input[:vin]
-      @status = input[:status] == nil ? :AVAILABLE : input[:status]
-
+      @status = input[:status] == nil ? :UNAVAILABLE : input[:status]
       @trips = input[:trips] == nil ? [] : input[:trips]
     end
 
+    def in_progress
+      return @trips.reject { |ride| ride.end_time == nil }
+    end
```
Comment (droberts-sea) [inline lib/driver.rb:27]: I really like the idea of this helper method. Great functional decomposition.

I'm not sure about the name, however. It looks like this method _rejects_ in-progress trips, so it would make sense to name it something like `completed_trips`. A method named `in_progress` I would expect to _only_ return trips that are in progress.

File: lib/driver.rb [hunk 2]
```diff
@@ -32,16 +37,45 @@ def average_rating
...
     def add_trip(trip)
-      if trip.class != Trip
+      unless trip.class <= Trip
         raise ArgumentError.new("Can only add trip instance to trip collection")
       end
-
+      # >>>>>Do i need this code below?<<<<<<
+      if trip.end_time.nil?
+        @status == :UNAVAILABLE
+      end
```
Comment (droberts-sea) [inline lib/driver.rb:50]: This looks good to me - if you add a trip and it's incomplete, you also modify the Driver's status.

File: lib/passenger.rb [hunk 3]
```diff
@@ -20,5 +23,27 @@ def get_drivers
     def add_trip(trip)
       @trips << trip
     end
+
+    def in_progress
+      return @trips.reject { |ride| ride.end_time == nil }
+    end
```
Comment (droberts-sea) [inline lib/passenger.rb:29]: You talked about this in your PR, but I don't like that this method is repeated for both `Driver` and `Passenger`. To me, it would be worthwhile to find a way to DRY this up.

File: lib/trip.rb [hunk 4]
```diff
@@ -10,12 +10,28 @@ def initialize(input)
...
-      @cost = input[:cost]
+      @cost = input[:cost] == nil ? 0 : input[:cost]
       @rating = input[:rating]
 
-      if @rating > 5 || @rating < 1
-        raise ArgumentError.new("Invalid rating #{@rating}")
+      if @end_time != nil && @rating != nil
+        if @end_time < @start_time
+          raise ArgumentError.new("Invalid date range")
+        elsif @rating > 5 || @rating < 1
```
Comment (droberts-sea) [inline lib/trip.rb:19]: The way this code is structured, if the end time is `nil` but the rating is not, the rating won't be checked.

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -75,27 +73,76 @@ def load_trips
...
       if id == nil || id <= 0
-        raise ArgumentError.new("ID cannot be blank or less than zero. (got #{id})")
+        raise ArgumentError.new("ID CANNOT BE BLANK OR LESS THAN ZERO (got #{id})")
       end
     end
+
+    def select_available_driver
+      @drivers.find do |driver|
+        if driver.status == :AVAILABLE
+          return driver
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:101]: Good work making this a separate method. This sort of thing would become especially valuable as you move onto wave 3.

File: lib/trip_dispatcher.rb [hunk 6]
```diff
@@ -75,27 +73,76 @@ def load_trips
...
+
+      new_ride[:start_time] = Time.now
+      new_ride[:end_time] = nil
+      new_ride[:cost] = nil
+      new_ride[:rating] = nil
+
+      active_trip = Trip.new(new_ride)
+
+      select_available_driver.add_trip(active_trip)
+      passenger.add_trip(active_trip)
```
Comment (droberts-sea) [inline lib/trip_dispatcher.rb:135]: Line 134 will call `select_available_driver` a second time! As your code is currently written this ends up working, but it's probably not what you want. Instead, you could say something like `active_trip.driver.add_trip(active_trip)`.

File: specs/driver_spec.rb [hunk 7]
```diff
@@ -12,7 +13,12 @@
     end
 
     it "throws an argument error with a bad ID value" do
-      proc{ RideShare::Driver.new(id: 0, name: "George", vin: "33133313331333133")}.must_raise ArgumentError
+      begin
+        RideShare::Driver.new(id: 0, name: "George", vin: "33133313331333133")
+        fail
```
Comment (droberts-sea) [inline specs/driver_spec.rb:18]: I like the exploration here, but in general it's preferable to use the built-in `must_raise` method.

File: specs/driver_spec.rb [hunk 8]
```diff
@@ -55,13 +63,27 @@
     end
   end
 
-  describe "average_rating method" do
-    before do
+  describe "Driver Status" do
+    it 'changes status if end_time equals nil' do
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
+      end_time = nil
```
Comment (droberts-sea) [inline specs/driver_spec.rb:69]: You should also check that it _doesn't_ change if you add a trip where the end time is non-nil.

File: specs/driver_spec.rb [hunk 9]
```diff
@@ -77,4 +99,93 @@
...
+
+    end
+    it 'calculate total revenue from driver' do
+      @driver.total_revenue.must_be_kind_of Float
+      @driver.total_revenue.must_equal 49.36
+    end
+    it 'calculate the average revenue per hour from driver' do
+      @driver.average_revenue.must_be_kind_of Float
+      @driver.average_revenue.must_equal 24.68
+    end
```
Comment (droberts-sea) [inline specs/driver_spec.rb:128]: You do a good job of catching edge cases here, especially for in-progress trips. One other you might ask is "what happens if the driver has no trips".

However, I think these tests could be better organized. Standard practice is to have one big outer `describe` block for the whole class, and then a sub-`describe` for each method. If there is some setup that's common to multiple sub-`describe`s, you can put it in a `before` block in the top-level `describe`. Something like this:

```ruby
describe 'Driver' do
  before do
    @driver_no_trips = Driver.new(...)
    @driver_complete_trips = Driver.new(...)
    # ... add a couple completed trips to @driver_complete_trips
    @driver_incomplete_trips = Driver.new(...)
    # ... add a couple incomplete trips
  end
  describe 'total_revenue' do
    it 'sums up trip revenues'
    it 'returns 0 if there are no trips'
    it 'ignores in-progress trips'
  end
end
```

File: specs/trip_dispatch_spec.rb [hunk 10]
```diff
@@ -89,4 +81,41 @@
...
+      # Act
+      driver = dispatcher.select_available_driver
+      # Assert
+      driver.id.must_equal 2
+      driver.status.must_equal :AVAILABLE
+    end
+  end
+
+  describe 'Request Trip' do
+    it 'will raise an ArgumentError when nothing is entered for Passenger ID' do
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:98]: What if there is no available driver? This is an important case, and is exactly the sort of situation where software will do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  @dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    @dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes - see responses above
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | yes - see inline comments
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | yes - good work
Complete tests | some missing - see inline
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** | Great job overall! Your code is clear and well-organized, you've done a great job of demonstrating functional decomposition, and your test coverage is coming along nicely. I've left a bunch of nitpicks below for you to review, but in general I'm quite pleased with this submission. I'm looking forward to seeing what you come up with on Hotel. Keep up the hard work!
